File: receiver_android/app/src/main/java/com/iflytek/lanmediacast/receiver/protocol/RemoteHttpHeaders.kt
package com.iflytek.lanmediacast.receiver.protocol

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

object RemoteHttpHeaders {
    const val MAX_COUNT = 5
    const val MAX_VALUE_BYTES = 2_048
    const val MAX_TOTAL_BYTES = 4_096

    private val allowedNames = mapOf(
        "user-agent" to "User-Agent",
        "referer" to "Referer",
        "origin" to "Origin",
        "accept" to "Accept",
        "accept-language" to "Accept-Language",
    )

    fun fromSource(source: JsonObject): Map<String, String> {
        val rawHeaders = source["httpHeaders"] ?: return emptyMap()
        val headers = rawHeaders as? JsonObject
            ?: throw IllegalArgumentException("httpHeaders must be an object")
        require(headers.size <= MAX_COUNT) { "too many remote HTTP headers" }

        val normalized = linkedMapOf<String, String>()
        var totalBytes = 0
        headers.forEach { (rawName, element) ->
            val name = allowedNames[rawName.lowercase()]
                ?: throw IllegalArgumentException("remote HTTP header is not allowed: $rawName")
            require(!normalized.containsKey(name)) { "duplicate remote HTTP header: $name" }
            val primitive = element as? JsonPrimitive
            require(primitive != null && primitive.isString) { "HTTP header values must be strings" }
            val value = primitive.content
            require(
                value.isNotEmpty() &&
                    value.trim() == value &&
                    value.all { character -> character.code in 0x20..0x7e },
            ) {
                "invalid remote HTTP header value: $name"
            }
            val valueBytes = value.toByteArray(Charsets.UTF_8).size
            require(valueBytes <= MAX_VALUE_BYTES) { "remote HTTP header value is too large: $name" }
            totalBytes += name.toByteArray(Charsets.UTF_8).size + valueBytes
            require(totalBytes <= MAX_TOTAL_BYTES) { "remote HTTP headers are too large" }
            normalized[name] = value
        }
        return normalized
    }
}

File: sender_flutter/lib/protocol/remote_http_headers.dart
import 'dart:convert';

const int maxRemoteHttpHeaders = 5;
const int maxRemoteHttpHeaderValueBytes = 2048;
const int maxRemoteHttpHeadersBytes = 4096;

const Map<String, String> remoteHttpHeaderNames = <String, String>{
  'user-agent': 'User-Agent',
  'referer': 'Referer',
  'origin': 'Origin',
  'accept': 'Accept',
  'accept-language': 'Accept-Language',
};

Map<String, String> validateRemoteHttpHeaders(Object? rawHeaders) {
  if (rawHeaders == null) return <String, String>{};
  if (rawHeaders is! Map) {
    throw const FormatException('httpHeaders must be an object');
  }
  if (rawHeaders.length > maxRemoteHttpHeaders) {
    throw const FormatException('too many remote HTTP headers');
  }

  final Map<String, String> normalized = <String, String>{};
  int totalBytes = 0;
  for (final MapEntry<Object?, Object?> entry in rawHeaders.entries) {
    final Object? rawName = entry.key;
    final Object? rawValue = entry.value;
    if (rawName is! String || rawValue is! String) {
      throw const FormatException(
        'HTTP header names and values must be strings',
      );
    }
    final String? name = remoteHttpHeaderNames[rawName.toLowerCase()];
    if (name == null) {
      throw FormatException('remote HTTP header is not allowed: $rawName');
    }
    if (normalized.containsKey(name)) {
      throw FormatException('duplicate remote HTTP header: $name');
    }
    if (rawValue.isEmpty ||
        rawValue.trim() != rawValue ||
        !rawValue.codeUnits.every(
          (int codeUnit) => codeUnit >= 0x20 && codeUnit <= 0x7e,
        )) {
      throw FormatException('invalid remote HTTP header value: $name');
    }
    final int valueBytes = utf8.encode(rawValue).length;
    if (valueBytes > maxRemoteHttpHeaderValueBytes) {
      throw FormatException('remote HTTP header value is too large: $name');
    }
    totalBytes += utf8.encode(name).length + valueBytes;
    if (totalBytes > maxRemoteHttpHeadersBytes) {
      throw const FormatException('remote HTTP headers are too large');
    }
    normalized[name] = rawValue;
  }
  return normalized;
}
